src/v4/options.rs

```diff
@@ -1368,6 +1523,36 @@ impl Encodable for DhcpOption {
                 }
                 encode_long_opt_bytes(code, &buf, e)?;
             }
+            ClasslessStaticRoute(routes) => {
```

Comment:
If the type is a `Vec<(Ipv4Addr, Ipv4Addr)>` then you can just add this to the `StaticRoutingTable` pattern match

Reply:
Oh, sorry. I glossed over the `Ipv4Net` and read it as `Ipv4Addr`, my bad!

The implementation of this should use the `encode_long_opt_chunks` function though. It will handle splitting the data sections for you, you just need to tell it how to write the items in the `Vec`

Reply:
Ha yes, the `Ipv4Net` requires a dedicated decode match arm.

I indeed glossed over `encode_long_opt_chunks`, I'll rewrite the encoding to use it. Thanks!

Reply:
On reading what `encode_long_opt_chunks` is doing, I don't think it'll be a perfect fit: a single route (`T` for the function) is variable in length, between 5 bytes for a /0 to 9 bytes for anything larger than /24. And reading the implementation in a more detailed way, it'll write an option with a len of `factor * chunk.len()`. This is wrong since any prefix length in `0..=24` will have an item size strictly smaller than `factor`.

To give an example, consider a single route: `172.16.0.0/12 via 10.0.0.1`. This route gets encoded as the following byte sequence: `12, 172, 16, 10, 0, 0, 1`. This is 7 bytes on the wire. Now if the table has a second route `192.168.10.0/24 via 10.0.0.1`, it gets encoded as `16, 192, 168, 10, 10, 0, 0, 1` which is 8 bytes. I think you see the problem here.

Reply:
Ah yeah, you're totally right. If we're okay with the bytes being split on any boundary, what do you think of this?

Reply:
It does indeed looks neat. Let me play around with it though, I don't know what would happend when splitting on byte boundary instead of route boundary. I'll test with a few DHCP clients to see how they handle it.

Reply:
So for now, I tested both `isc-dhcp-client` and `dhcpcd` and both accept properly options split at byte boundaries (not route boudaries). I'll test systemd-networkd and NetworkManager now.

One thing that does not inspire confidence is that it does break Wireshark. Decoding of the DHCP packet stops at the end of the option with a "Malformed Packet" error.

Reply:
Interesting, thanks for testing. I am almost certain that all implementations of long form encoding just split at the 255 byte boundary and not where each section needs to be decodable individually.

According to the RFC:
>  Options can be split on any octet boundary.  No split portion of an
   option that has been split can contain more than 255 octets.  The
   split portions of the option MUST be stored in the aggregate option
   buffer in sequential order - the first split portion MUST be stored
   first in the aggregate option buffer, then the second portion, and so
   on.  The encoding agent MUST NOT attempt to specify any semantic
   information based on how the option is split.

https://datatracker.ietf.org/doc/html/rfc3396

What did you use exactly to test in wireshark? If I could replicate that here I may be able to look into it.

We are using that same long encoding style for many different options in the parser so it would be good to figure this out.

Reply:
I would bet wireshark just doesn't support long option encoding. I'm fairly certain some DHCP servers like Kea don't support it either. I'll see if I can find a ticket on wiresharks gitlab page...

I'm okay with merging this after we resolve the other issue though. I can look into wireshark without holding this ticket up. I think I'll try a few different options encoded with the long format and see how it fares.

Thank you for all your effort on this PR!

Off-topic, do you have an email I can contact you with? If you don't want to post publicly, you can get me at ecameron@bluecatnetworks.com

Reply:
Yup I do agree this looks more like a Wireshark issue than dhcproto's.

Sure I do, depending on which one you want :smile: (nickname, personal, professional).